src/qibolab/instruments/zhinst.py

```diff
@@ -0,0 +1,623 @@
+
+        self.pulse = pulse
+        self.signal = f"{pulse.type.name.lower()}{pulse.qubit}"
+        self.zhpulse = ZhPulse(pulse).zhpulse
+
+        # Stores the baking object (for pulses that need 1ns resolution)
+        self.shot = None
+        self.IQ = None
+        self.shots = None
+        self.threshold = None
+
+    def select_sweeper(self, sweeper, parameter):
+        if parameter == "amplitude":
+            Zh_Sweeper = lo.SweepParameter(
+                uid=sweeper.parameter.name,
+                values=sweeper.values,
+            )
+        elif parameter == "length":
+            Zh_Sweeper = lo.SweepParameter(
+                uid=sweeper.parameter.name,
+                values=sweeper.values * 10e-9,
+            )
+        elif parameter == "frequency":
+            Zh_Sweeper = lo.SweepParameter(
+                uid=sweeper.parameter.name,
+                values=sweeper.values,
+            )
+        elif parameter == "flux_bias":
+            Zh_Sweeper = lo.SweepParameter(
+                uid=sweeper.parameter.name,
+                values=sweeper.values,
+            )
+        return Zh_Sweeper
+
+
+class Zurich(AbstractInstrument):
+    def __init__(self, name, descriptor, use_emulation=False):
+        self.name = name
+        self.descriptor = descriptor
+        self.emulation = use_emulation
+
+        self.is_connected = False
+
+        self.time_of_flight = 0
+        self.smearing = 0
+
+    def connect(self):
+        if not self.is_connected:
+            for attempt in range(3):
+                try:
+                    self.device_setup = lo.DeviceSetup.from_descriptor(
+                        yaml_text=self.descriptor,
+                        server_host="localhost",
+                        server_port="8004",
+                        setup_name=self.name,
+                    )
+                    self.session = lo.Session(self.device_setup)
+                    self.device = self.session.connect(do_emulation=self.emulation)
+                    # self.device.reset()
+                    self.is_connected = True
+                    break
+                except Exception as exc:
+                    print(f"Unable to connect:\n{str(exc)}\nRetrying...")
+            if not self.is_connected:
+                raise InstrumentException(self, f"Unable to connect to {self.name}")
+
+    def start(self):
+        pass
+
+    def stop(self):
+        pass
+
+    def disconnect(self):
+        if self.is_connected:
+            self.device = self.session.disconnect()
+            self.is_connected = False
+        else:
+            print(f"Already disconnected")
+
+    def setup(self, qubits, relaxation_time=0, time_of_flight=0, smearing=0, **_kwargs):
+        self.relaxation_time = relaxation_time
+        self.time_of_flight = time_of_flight
+        self.smearing = smearing
+
+        self.signal_map = {}
+        self.calibration = lo.Calibration()
```

Comment:
> This may just be my personal opinion, but I generally try to define all class attributes in `__init__`

Not a personal opinion, or at least not only yours: PyLint enforces this with [W0201](https://pylint.readthedocs.io/en/latest/user_guide/messages/warning/attribute-defined-outside-init.html)